src/tss2-mu/tpms-types.c

```diff
@@ -39,210 +39,129 @@
 #define VAL
 #define TAB_SIZE(tab) (sizeof(tab) / sizeof(tab[0]))
 
-static TSS2_RC marshal_pcr_select(const UINT8 *ptr, uint8_t buffer[],
-                                  size_t buffer_size, size_t *offset)
-{
-    TPMS_PCR_SELECT *pcrSelect = (TPMS_PCR_SELECT *)ptr;
-    UINT32 i;
-    TSS2_RC ret;
-
-    if (!ptr) {
-        LOG_WARNING("src param is NULL");
-        return TSS2_MU_RC_BAD_REFERENCE;
-    }
-
-    ret = Tss2_MU_UINT8_Marshal(pcrSelect->sizeofSelect, buffer, buffer_size, offset);
-    if (ret)
-        return ret;
-
-    if (pcrSelect->sizeofSelect > TAB_SIZE(pcrSelect->pcrSelect)) {
-        LOG_ERROR("sizeofSelect value too big");
-        return TSS2_SYS_RC_BAD_VALUE;
-    }
-
-    for (i = 0; i < pcrSelect->sizeofSelect; i++)
-    {
-        ret = Tss2_MU_UINT8_Marshal(pcrSelect->pcrSelect[i], buffer, buffer_size, offset);
-        if (ret)
-            return ret;
-    }
-
-    return TSS2_RC_SUCCESS;
-}
-
-static TSS2_RC unmarshal_pcr_select(uint8_t const buffer[], size_t buffer_size,
-                                    size_t *offset, UINT8 *ptr)
-{
-    TPMS_PCR_SELECT *pcrSelect = (TPMS_PCR_SELECT *)ptr;
-    UINT32 i;
-    TSS2_RC ret;
-
-    if (!ptr) {
-        LOG_WARNING("dest param is NULL");
-        return TSS2_MU_RC_BAD_REFERENCE;
-    }
-
-    ret = Tss2_MU_UINT8_Unmarshal(buffer, buffer_size, offset, &pcrSelect->sizeofSelect);
-    if (ret)
-        return ret;
-
-    if (pcrSelect->sizeofSelect > TAB_SIZE(pcrSelect->pcrSelect)) {
-        LOG_ERROR("sizeofSelect value too big");
-        return TSS2_SYS_RC_MALFORMED_RESPONSE;
-    }
-
-    for (i = 0; i < pcrSelect->sizeofSelect; i++)
-    {
-        ret = Tss2_MU_UINT8_Unmarshal(buffer, buffer_size, offset, &pcrSelect->pcrSelect[i]);
-
-        if (ret)
-            return ret;
-    }
-
-    return TSS2_RC_SUCCESS;
+#define TPMS_PCR_MARSHAL(type, before) \
+TSS2_RC \
+Tss2_MU_##type##_Marshal(const type *src, uint8_t buffer[], \
+                           size_t buffer_size, size_t *offset) \
+{ \
+    TSS2_RC ret = TSS2_RC_SUCCESS; \
+    size_t local_offset = 0; \
+    UINT8 i; \
+\
+    if (!src) { \
+        LOG_WARNING("src param is NULL"); \
+        return TSS2_MU_RC_BAD_REFERENCE; \
+    } \
+\
+    if (offset) { \
+        local_offset = *offset; \
+    } else if (!buffer) { \
+        return TSS2_MU_RC_BAD_REFERENCE; \
+    } \
+\
+    LOG_DEBUG( \
+         "Marshalling " #type " from 0x%" PRIxPTR " to buffer 0x%" PRIxPTR \
+         " at index 0x%zx", (uintptr_t)&src, (uintptr_t)buffer, \
+         offset?*offset:0xffff); \
+\
+    if (src->sizeofSelect > TAB_SIZE(src->pcrSelect)) { \
+        LOG_ERROR("sizeofSelect value %"PRIu8"/%zi too big", src->sizeofSelect, \
+                  TAB_SIZE(src->pcrSelect)); \
+        return TSS2_SYS_RC_BAD_VALUE; \
+    } \
+\
+    before \
```

Comment:
This "before" looks weird. Can we change it to something that will look more like a macro:
```
#define TPMS_PCR_MARSHAL(type, tag) \
                                  size_t buffer_size, size_t *offset)	+TSS2_RC \
{	+Tss2_MU_##type##_Marshal(const type *src, uint8_t buffer[], \
    TPMS_PCR_SELECT *pcrSelect = (TPMS_PCR_SELECT *)ptr;
                         size_t buffer_size, size_t *offset) \
{ \
....
ret = tag; \
...
}\
```
and then:
```
TPMS_PCR_MARSHAL(TPMS_PCR_SELECT,  TSS2_RC_SUCCESS)
TPMS_PCR_MARSHAL(TPMS_PCR_SELECTION, \
Tss2_MU_TPMI_ALG_HASH_Marshal(src->hash, buffer, buffer_size, &local_offset))
TPMS_PCR_MARSHAL(TPMS_TAGGED_PCR_SELECT, \
Tss2_MU_UINT32_Marshal(src->tag, buffer, buffer_size, &local_offset))
```

Reply:
sure... we can do that...